File: android/src/main/kotlin/io/polyfence/core/ZoneConfig.kt
package io.polyfence.core

/**
 * Typed configuration for creating geofence zones.
 * Preferred over the raw Map-based addZone API.
 */
data class ZoneConfig(
    val id: String,
    val name: String,
    val type: GeofenceEngine.ZoneType,
    val center: GeofenceEngine.LatLng? = null,
    val radius: Double? = null,
    val polygon: List<GeofenceEngine.LatLng>? = null,
    val metadata: Map<String, String>? = null
) {
    companion object {
        /**
         * Create a circle zone configuration
         */
        fun circle(
            id: String,
            name: String,
            center: GeofenceEngine.LatLng,
            radius: Double,
            metadata: Map<String, String>? = null
        ): ZoneConfig {
            return ZoneConfig(
                id = id,
                name = name,
                type = GeofenceEngine.ZoneType.CIRCLE,
                center = center,
                radius = radius,
                metadata = metadata
            )
        }

        /**
         * Create a polygon zone configuration
         */
        fun polygon(
            id: String,
            name: String,
            polygon: List<GeofenceEngine.LatLng>,
            metadata: Map<String, String>? = null
        ): ZoneConfig {
            return ZoneConfig(
                id = id,
                name = name,
                type = GeofenceEngine.ZoneType.POLYGON,
                polygon = polygon,
                metadata = metadata
            )
        }
    }

    /**
     * Convert to the Map format used by the legacy addZone API
     */
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        when (type) {
            GeofenceEngine.ZoneType.CIRCLE -> {
                map["type"] = "circle"
                center?.let {
                    map["center"] = mapOf("latitude" to it.latitude, "longitude" to it.longitude)
                }
                radius?.let { map["radius"] = it }
            }
            GeofenceEngine.ZoneType.POLYGON -> {
                map["type"] = "polygon"
                polygon?.let { points ->
                    map["polygon"] = points.map { mapOf("latitude" to it.latitude, "longitude" to it.longitude) }
                }
            }
        }
        metadata?.let { map["metadata"] = it }
        return map
    }
}

File: ios/Classes/ZoneConfig.swift
import Foundation
import CoreLocation

/**
 * Typed configuration for creating geofence zones.
 * Preferred over the raw Dictionary-based addZone API.
 */
public struct ZoneConfig {
    public let id: String
    public let name: String
    public let type: ZoneType
    public let center: CLLocationCoordinate2D?
    public let radius: Double?
    public let polygon: [CLLocationCoordinate2D]?
    public let metadata: [String: String]?

    /**
     * Create a circle zone configuration
     */
    public static func circle(
        id: String,
        name: String,
        center: CLLocationCoordinate2D,
        radius: Double,
        metadata: [String: String]? = nil
    ) -> ZoneConfig {
        return ZoneConfig(
            id: id,
            name: name,
            type: .circle,
            center: center,
            radius: radius,
            polygon: nil,
            metadata: metadata
        )
    }

    /**
     * Create a polygon zone configuration
     */
    public static func polygon(
        id: String,
        name: String,
        polygon: [CLLocationCoordinate2D],
        metadata: [String: String]? = nil
    ) -> ZoneConfig {
        return ZoneConfig(
            id: id,
            name: name,
            type: .polygon,
            center: nil,
            radius: nil,
            polygon: polygon,
            metadata: metadata
        )
    }

    /**
     * Convert to the Dictionary format used by the legacy addZone API
     */
    public func toMap() -> [String: Any] {
        var map: [String: Any] = [:]
        switch type {
        case .circle:
            map["type"] = "circle"
            if let center = center {
                map["center"] = ["latitude": center.latitude, "longitude": center.longitude]
            }
            if let radius = radius {
                map["radius"] = radius
            }
        case .polygon:
            map["type"] = "polygon"
            if let polygon = polygon {
                map["polygon"] = polygon.map { ["latitude": $0.latitude, "longitude": $0.longitude] }
            }
        }
        if let metadata = metadata {
            map["metadata"] = metadata
        }
        return map
    }
}
